Refactor zabbix server

eb4x opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR consists of a lot of iterative commits, they should all be (somewhat) self-explanatory, doing a "single self-contained" change and a rationale for it. If one of the changes are not to anyones liking, comment on it and we can either improve upon it or take it out.

This is a considerably big refactor of the zabbix_server role. I'm not expecting it to break for existing installations and whatever overrides have been provided. It's not intended to "fix" any issues, just clean up the tasks and make the experience somewhat more consistent. (Though it does bring back support for deploying Zabbix 5.0 LTS.)

The most "controversial" change might be stripping out the zabbix_valid_server_versions check.

Other decisions include moving package installation out of the os_family tasks, tasked with setting up the repo. But they did much more beyond that. So the installation happens in main.yml, after repo setup, and the dependencies for db-variants are installed in the mysql/postgresql task files, (instead of checking which dependencies are needed for which db-variant for which os, during repo setup)

Arguably the repo-setup should be it's own role, as it's repeated out for every single role.

##### ISSUE TYPE
- Refactor Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
zabbix_server

##### ADDITIONAL INFORMATION
I'd like for this refactor to retain all its commits and messages, and not become a squash merge. One thing I sorely missed during this refactor was the ability to find a log as to why things were done in a certain way.

I couldn't quite get molecule to work by itself, and to iterate and test multiple platforms I modified it a bit to have things happen in parallel. I also got a dedicated machine with multiple gh-runners set-up, but can't really explain why wasn't a great success with the existing workflow.

Also, I couldn't get the geerlingguy images to work, but I think the thing we want are simple systemd based images (alma and rocky provide these already) with python3 installed, which turns out being very simple Dockerfiles to create.
```
FROM {{ item.base_image }}
RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y python3-pip systemd-sysv
CMD ["/sbin/init"]
```
So maybe rely on upstream images instead of geerlingguy in the future?

You can check out these kinds of changes here; https://github.com/eb4x/community.zabbix/blob/molecule5/molecule/zabbix_server/molecule.yml

eb4x commented on roles/zabbix_server/vars/Debian.yml:
```diff
@@ -6,7 +6,6 @@ apache_log: apache2
 debian_keyring_path: /etc/apt/keyrings/
-zabbix_gpg_key: "{{ debian_keyring_path }}zabbix-repo.asc"
-_zabbix_repo_deb_url: "http://repo.zabbix.com/zabbix/{{ zabbix_server_version }}"
+zabbix_gpg_key: "{{ debian_keyring_path }}/zabbix-official-repo.asc"
```
This is change is a mistake introduced during rebasing. Will fix.

eb4x commented on roles/zabbix_server/tasks/mysql.yml:
```diff
@@ -49,170 +49,102 @@
+        collation: "{{ zabbix_server_dbcollation }}"
+        state: import
+        target: "{{ is_legacy_version | ternary(legacy_path, modern_path) }}"
+        use_shell: true
```
Also, I think use_shell is not needed. Will test before I update the pr.

pyrodie18 commented:
So without having seen if molecule pushes through all of this (just hit the go button) I will say this.  First, you did a ton of work here, so thank you for that.  It is a big refactor.  I will say some of the reason why tasks are broken out the way they are (in Debian vs main for example) most likely goes back to how things were put together < 2.0 when we were trying to accommodate pretty much every version of Zabbix known to man.  We cleaned up a lot of it, but kept the same basic form-factor of task distribution.  I'm not opposed to doing that, but if we are, I would only support it if it was a common refactor across all of the roles.

I don't personally support bringing back 5.0.  One if we're going to do it, we need to do it for all of the roles but more importantly our current stance is we only support fully supported versions of Zabbix.  Yes it limits some people's use cases but the goal of the collection is not to provide 100% support for every use case, but to support the most common.  

In one of your PRs you made a comment about the user should know the risks of trying to run against an unsupported version.  I agree that they should but when you look at some of the tickets that get created, its obvious that they don't which is why we installed the sanity check.  My own personal thinking when I introduced that was if they understood the risks, then they would also be able to comment that task out easy enough to do whatever they were wanting to do.

I'm curious to hear @D3DeFi  and @BGmot  thoughts.

dj-wasabi commented:
It is a huge effort made by you @eb4x , but I totally support @pyrodie18 with Zabbix 5.0 LTS, it is end of support since last year may (https://www.zabbix.com/life_cycle_and_release_policy). We can not keeping old versions like the beginning of the role when it was under my name, it takes a lot of effort to maintain them.

eb4x commented:
Thank you for taking a look :smiley:

> We cleaned up a lot of it, but kept the same basic form-factor of task distribution. I'm not opposed to doing that, but if we are, I would only support it if it was a common refactor across all of the roles.

I just wanted to get this PR out the door to see how it would be received. If this is something we would go forward with, I'll do a refactor on the rest of the roles, and this PR can just sit and wait.

> I don't personally support bringing back 5.0.

There was so much code in place to bring it back already, it was like 8 lines missing. With the amount of code-reduction in this refactor it was a fun and easy thing to add.

> we need to do it for all of the roles but ...

Can do :smile_cat: !

> but when you look at some of the tickets that get created, its obvious that they don't which is why we installed the sanity check.

Fair point :+1: 

> they would also be able to comment that task out easy enough to do whatever they were wanting to do.

Perhaps supply a flag to a when condition of such a task, so we can bypass it. `-e skip_version_check=true`? I'll refactor in that change.

Zabbix 7 is coming soon. I've already done some work to the plugins to support the renamed mappings, but only for the ones I needed. That codebase could use a good refactor aswell, but I'd need some guidance and hear some opinions or preferences for the code.

pyrodie18 commented:
> I just wanted to get this PR out the door to see how it would be received. If this is something we would go forward with, I'll do a refactor on the rest of the roles, and this PR can just sit and wait.
Let me download this PR and look at it without the comparisons (its really hard to see simplification with all of the red lines but I'm sure its there and give you a better opinion later on today or this weekend.

> There was so much code in place to bring it back already, it was like 8 lines missing. With the amount of code-reduction in this refactor it was a fun and easy thing to add.
I agree but then it becomes us going down this road that doesn't end.  If I remember correctly there we comments made about us dropping support for 4.0 even though it was dead dead by that point and I'm sure the same argument could be made about only 9 lines (no idea but its a guess).

> Perhaps supply a flag to a when condition of such a task, so we can bypass it. `-e skip_version_check=true`? I'll refactor in that change.
Ya I could support that

pyrodie18 commented:
One other comment I will make about bringing back 5.0 is it grows the testing matrix substantially to add another version.  It already takes something like 2 hours to run through the full test matrix when we affect every role and plugin.  Also we removed all of the 5.0 support from the modules which is definitely more than 9 lines.

eb4x commented:
> One other comment I will make about bringing back 5.0 is it grows the testing matrix substantially to add another version. It already takes something like 2 hours to run through the full test matrix when we affect every role and plugin. 

Well, atleast we can strip out 6.2 now right? But yes, no-one in their right mind should be deploying 5.0, so I'm fully behind not testing for it. Since the scripts had a check for whether to use create.sql or server.sql, I opted to complete that functionaliy instead of taking it away. There is very little else required, just some config-options. (How about we not tell anyone the roles can also deploy 5.0 again?)

> Also we removed all of the 5.0 support from the modules which is definitely more than 9 lines.

Ah, didn't think about those.

eb4x commented:
Ok, so I took a 32nd look at this code, and there is a flaw here. It was always there, but the ansible python database dependencies are installed on the zabbix_server even though the database creation tasks could be delegated to an external database, in which case the dependencies could be missing. This was a problem before the refactor, it remains a problem after.

I'll drum up a fix for that.

pyrodie18 commented:
> Well, atleast we can strip out 6.2 now right?

Removing support for a already working feature/version is a major change.  When we release 3.0 (which we plan on doing with the release of Zabbix 7) we'll drop support for 6.2 and 6.4 as well as a few operating systems.

eb4x commented on roles/zabbix_server/tasks/Debian.yml:
```diff
@@ -129,116 +95,19 @@
-
-- name: "Debian | Allow Zabbix dpkg installs to /usr/share/doc/zabbix*"
+# Since this is where Zabbix 5.0 installs its database schemas, we need to allow
+# files to be installed to /usr/share/doc/zabbix-server-*
+- name: "Debian | Allow Zabbix dpkg installs to /usr/share/doc/zabbix-server-*"
   ansible.builtin.lineinfile:
     path: /etc/dpkg/dpkg.cfg.d/excludes
-    line: "path-include=/usr/share/doc/zabbix*"
+    create: true
+    line: 'path-include=/usr/share/doc/zabbix\-server\-*'
```
Yeah, I discovered a similar task in molecule/zabbix_proxy/prepare.yml, and I think I questioned (in this commit) whether or not this should be done there instead. I've since moved it there. But yeah, it should get removed eventually.

eb4x replied:
Done in 3923029571441fa265358fe1384617a2bb70be08

eb4x commented on roles/zabbix_server/tasks/initialize-mysql.yml:
```diff
@@ -0,0 +1,149 @@
+    - dependencies
+
+- name: "MySQL Database prep"
+  when: zabbix_server_database_creation
+  delegate_to: "{{ zabbix_server_real_dbhost | default(zabbix_server_dbhost_run_install | ternary(delegated_dbhost, inventory_hostname)) }}"
+  vars:
+    delegated_dbhost: "{{ (zabbix_server_dbhost == 'localhost') | ternary(inventory_hostname, zabbix_server_dbhost) }}"
+  tags:
+    - database
```
Yes, until I started using ansible-lint, I also always put the `when`, `delegate_to`, `vars` and such at the bottom. ansible-lint had a pretty good argument for putting these things before `block/rescue/always`. Since then (even though it was a bit jarring at first) I've started putting `when` right after `name` for many non-block tasks aswell. It forces you to write a better description for `name` and you can immediately read the condition.

> "Initialize database, when zabbix_server_database_creation

or

> "Install packages, when item.required"

It doesn't feel right for every type of tast though. And again, I know it looks and feels weird at first, but if you give it a chance, it'll grow on you :smile_cat:

eb4x commented on roles/zabbix_server/tasks/initialize-mysql.yml:
```diff
@@ -0,0 +1,149 @@
+          register: _mysql_variable_defaults
+
+        - name: "MySQL | Set variable overrides for schema import"
+          when: item.msg != _mysql_schema_import_overrides[item.name]
```
Yep, kinda agree. Will update

eb4x replied:
Done.

eb4x commented on roles/zabbix_server/tasks/initialize-pgsql.yml:
```diff
@@ -1,126 +1,96 @@
+        state: restore
+        target: "{{ is_legacy_version | ternary(legacy_path, modern_path) }}"
+      vars:
+        is_legacy_version: "{{ zabbix_server_version is version('6.0', '<') }}"
```
Yeah, that's what I meant when I said most of the code for deploying 5.0 was still there. I'd rather not break anything just yet. It's kinda fun for me refactoring, but maintaining all functionality.

There's already so much code trimmed, and the 5.0 tasks are easily identifiable.

Will eventually create a commit that strips it out though, don't worry :)

eb4x commented:
Apologies for the hold up, been busy trying to get my own actions-runner working so I can get these results myself, as currently I'm currently just testing against [a few select platforms](https://github.com/eb4x/community.zabbix/blob/molecule5/molecule/zabbix_server/molecule.yml#L4-L164) and using images generated from Dockerfiles.

I'm currently experimenting with running a [kind](https://kind.sigs.k8s.io/) kubernetes cluster, and the [actions-runner-controller](https://github.com/actions/actions-runner-controller/), and had some marginal success getting it to run atleast :sweat_smile:

I'll get back to this once I have gotten repeatable tests.

eb4x commented on roles/zabbix_server/tasks/Debian.yml:
```diff
@@ -3,40 +3,6 @@
-  tags:
-    - always
-
-- name: "Debian | Installing lsb-release"
-  ansible.builtin.apt:
-    pkg: lsb-release
-    update_cache: true
-    cache_valid_time: 3600
-    force: true
-    state: present
-  environment:
-    http_proxy: "{{ zabbix_http_proxy | default(None) | default(omit) }}"
-    https_proxy: "{{ zabbix_https_proxy | default(None) | default(omit) }}"
-  become: true
-  tags:
-    - install
-
-- name: "Debian | Update ansible_lsb fact"
-  ansible.builtin.setup:
-    gather_subset:
-      - lsb
-
```
I've updated the code to work with Raspbian, and made a note of my findings in 5c6337a6cf12bf69a1e39da1d3c3cc2b951e8716

eb4x commented on roles/zabbix_server/templates/zabbix_server.conf.j2:
Done in 3923029571441fa265358fe1384617a2bb70be08

eb4x commented:
Apologies for the hold-up. To borrow a sentence from @fasterthanlime

> GitHub actions-runners do not spark joy.

So there are the important changes since last;
- using pymysql for all platforms (swapping out mysqldb on Debian systems, reasoning is explained in 02102f3afa37fc265159d6eff97cf7bd00b335fc)
- bringing back lsb to identify raspbian (details in 5c6337a6cf12bf69a1e39da1d3c3cc2b951e8716)
- dropping code that made zabbix 5.0 installations work (2fa5fc6ad38d019d9c7b30d4c8864b1c5e961860)

(I'll create a separate PR to address some issues I had with scaling actions-runner)

eb4x commented:
Just adding a note here as I took a look at cleaning up the selinux.yml tasks.

I think the inclusion of selinux.yml needs to move into main.yml after the zabbix packages are installed for some of the sebooleans to exist. I'll create a Vagrantfile and do some testing on vm's as the molecule containers won't test the selinux bits.

eb4x commented:
> I think the inclusion of selinux.yml needs to move into main.yml after the zabbix packages are installed for some of the sebooleans to exist.

I was completely wrong about this, the booleans are present on systems that have no zabbix software.

pyrodie18 commented on roles/zabbix_server/tasks/main.yml:
```diff
@@ -31,13 +31,18 @@
+  when: item.required | default(true)
+  loop:
+    - name: "{{ zabbix_server_package | default(_zabbix_server_package) }}"
+    - required: "{{ zabbix_server_version is version('6.0', '>=') }}"
```
OK maybe you can help me understand what you're doing here.  What is the required parameter doing?  I don't actually see it being used anywhere?

eb4x replied:
It's getting used in the when condition above. (Though all this code has since been removed with the 5.0 removal)
